adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,285 @@
+	for i := 0; i < len(request.Imp); i++ {
+		imp := &request.Imp[i]
+
+		var ext unicornImpExt
+		err := json.Unmarshal(imp.Ext, &ext)
+
+		if err != nil {
+			return nil, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while decoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		var placementID string
+		if ext.Bidder.PlacementID != "" {
+			placementID = ext.Bidder.PlacementID
+		} else {
+			placementID, err = getStoredRequestImpID(imp)
+			if err != nil {
+				return nil, &errortypes.BadInput{
+					Message: fmt.Sprintf("Error get StoredRequestImpID from imp[%d]: %s", i, err),
+				}
+			}
+		}
+
+		ext.Bidder.PlacementID = placementID
+
+		imp.Ext, err = json.Marshal(ext)
+		if err != nil {
+			return nil, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while encoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		secure := int8(1)
+		imp.Secure = &secure
+		imp.TagID = placementID
+	}
+	return request.Imp, nil
+}
+
+func getStoredRequestImpID(imp *openrtb.Imp) (string, error) {
+	var impExt map[string]json.RawMessage
+
+	err := json.Unmarshal(imp.Ext, &impExt)
+
+	if err != nil {
+		return "", fmt.Errorf("Error while decoding ext because: %s", err)
+	}
+
+	rawPrebidExt, ok := impExt[openrtb_ext.PrebidExtKey]
+
+	if !ok {
+		return "", fmt.Errorf("ext.prebid is null")
+	}
+
+	var prebidExt openrtb_ext.ExtImpPrebid
+
+	err = json.Unmarshal(rawPrebidExt, &prebidExt)
+
+	if err != nil {
+		return "", fmt.Errorf("cannot decoding ext.prebid because: %s", err)
+	}
+
+	if prebidExt.StoredRequest == nil {
+		return "", fmt.Errorf("ext.prebid.storedrequest is null")
+	}
+
+	return prebidExt.StoredRequest.ID, nil
+}
+
+func setSourceExt() (json.RawMessage, error) {
+	siteExt, err := json.Marshal(unicornSourceExt{
+		Stype:  "prebid_server_uncn",
+		Bidder: "unicorn",
+	})
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error while encoding source.ext, err: %s", err),
+		}
+	}
+	return siteExt, nil
+}
+
+func setExt(request *openrtb.BidRequest) (json.RawMessage, error) {
+	accountID, err := jsonparser.GetInt(request.Imp[0].Ext, "bidder", "accountId")
+	if err != nil {
+		accountID = 0
+	}
+	var decodedExt *unicornExt
+	err = json.Unmarshal(request.Ext, &decodedExt)
+	if err != nil {
+		decodedExt = &unicornExt{
+			Prebid: nil,
+		}
+	}
+	decodedExt.AccountID = accountID
+
+	ext, err := json.Marshal(decodedExt)
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error while encoding ext, err: %s", err),
+		}
+	}
+	return ext, nil
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *UnicornAdapter) MakeBids(request *openrtb.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
```

Comment:
Please add test coverage for this condition.

Reply:
I mistaked on 400.json. Fixed.